### ISSUE TITLE
PluginServiceProvider's subproviders are unable to access eachother's services when activating a service

### ISSUE THREAD

**Issue by snixtho** [EvoTM/EvoSC-sharp#46]
# The bug
There appear to be an issue when activating services that depend on services from other service providers.

For example, if a service, `Plugin1Service`, from `Plugin1` depends on a service, `Plugin2Service`, that is registered from `Plugin2`. The plugin system will create a `PluginServiceProvider` that contains the core's service provider, `Plugin1`'s service provider, which contains `Plugin1Service` and `Plugin2`'s service provider, which contains `Plugin2Service`. And, `Plugin1Service` uses the DI system, and gets `Plugin2Service` passed in it's constructor.

Now, imagine you have a command that a user calls. This command requires `Plugin1Service` and uses DI in it's constructor to get this service. When the command is called, the DI system will attempt to activate `Plugin1Service`. However, since `Plugin2Service` does not exist in `Plugin1`'s service provider, the activation fails.

# The problem
The problem here is that the different service providers within the same `PluginServiceProvider` has no knowledge of any of each other's services.

To visualize the problem here is a diagram:
![Untitled Diagram drawio (1)](https://user-images.githubusercontent.com/3867739/192457536-491eda65-7d25-438f-aa2d-bbea7bd24540.png)

Because the way service providers works, there is no way to provide this knowledge dynamically during runtime after the ServiceProvider has been built.

# Consequences
What this means is that it breaks the entire dependency system. Plugins have no way to communicate with eachother, access eachother's services. The problem becomes even bigger because a plugin's services can't even access the core services (eg. no database access, rpc access etc.)!

**Comment by janmarques** [comment]
what is the desired design for managing dependencies? Should plugins be only allowed to access the root system + some other plugins that the are dependant on, or is it not that important?

If it is not important, the simplest solution may include all the plugin dependencies in the IServiceCollection of the root system, and then instantiate the plugin from the root system as needed with https://learn.microsoft.com/en-us/dotnet/api/microsoft.extensions.dependencyinjection.serviceproviderserviceextensions.createscope?view=dotnet-plat-ext-6.0 in the callback hooks that we want?

**Comment by AtomicLiquid** [comment]
The initial plan was to try and figure out if the plugins could register their dependencies within the existing ServiceProvider within the core. That way we could keep track of which plugins were in charge of what action based on their registered services. However we faced problems with the Service Provider as it had been initialized during startup. When the ServiceProvider have been initialized, it is not possible to register new services or remove existing services as it's not the intended use based on Microsofts DI framework's specifications. Because of this, we had to rethink the architecture, and also how the plugins should be able to communicate with the core and how the plugins should be able to communicate with eachother. 

Our current plan is to not allow the plugins to register their services within the core's ServiceProvider. They will still be able to access the core's services. If they want to register new services, they have to create a new ServiceProvider which has to be isolated from the core and scoped to the plugin itself. The plugins will then communicate with the core using events and registries (collections).

So, to sum it up. The desired design is currently a WIP, plugins can communicate with eachother, but have to go through the core (root system). And last but not least, it's not possible to add/remove plugins after the ServiceProvider has been initialized, and as the plugins can be enabled and disabled during runtime, the solution you propose won't work unfortunately.

**Comment by janmarques** [comment]
OK so more like a good old MAF/MEF system then, running with a little more isolation and restriction. Good luck 👍
